### ISSUE TITLE
Overly verbose descriptions for interface methods

### ISSUE THREAD

**Issue by adamschoenemann** [idris-lang/Idris-dev#3268]
I understand the fact that informative messages are useful, but I maybe think that the verbosity involved when disambiguating interface methods is a bit too much. For example, I recently got this output when inspecting the type of a hole:

```idris

  a : Type
  x : a
  w : Nat
  k : Nat
  c : Vect k a
  m : Vect w (Vect (S k) a)
  induction : vtrans (Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                             m) =
              Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                     (vtrans (c ::
                                                                                              Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                                                                                                     m))
  u : Type
  lem1' : Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map head
                                                                                 m ::
          vtrans (Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                         m) =
          Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map head m ::
          Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                 (vtrans (c ::
                                                                                          Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                                                                                                 m))
--------------------------------------
rhs : vtrans m =
      Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map head m ::
      Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                             (vtrans (c ::
                                                                                      Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map tail
                                                                                                                                                             m))
```
At this point I basically gave up attempting to glean any useful information from this output. Maybe a quick and dirty solution would be to alias the methods and then write the "alias definition" at the bottom (or top), e.g. rename `map` to `vect_map` or something else and then write:

"where `vect_map` is the Data.Vect.Vect n implementation of Prelude.Functor.Functor, method map"

at e.g. the bottom.
Maybe I am the only one being annoyed by this? Maybe I am doing something wrong, I am still a very much a newb at Idris.

Thanks

**Comment by jfdm** [comment]
@adamschoenemann Thanks for the issue. I can understand that such signatures are not intuitive nor helpful.

What would be of help here would be a a Minimum Working Example to use when addressing the issue.

Thanks.

**Comment by adamschoenemann** [comment]
Of course, here is one :)

```idris
example : {xs : List t} -> map f (map g xs) = map (f . g) xs
example = ?example_rhs
```
This seemingly simple signature gives this whopping output when loaded into the repl and inspected with `:t`

```idris
  b : Type
  a : Type
  f : a -> b
  t : Type
  g : t -> a
  xs : List t
--------------------------------------
Main.example_rhs : Prelude.List.List implementation of Prelude.Functor.Functor, method map f
                                                                                              (Prelude.List.List implementation of Prelude.Functor.Functor, method map g

                                                              xs) =
                      Prelude.List.List implementation of Prelude.Functor.Functor, method map (\x =>
                                                                                                 f (g x))

                                                                                              xs
```
What you'd want is (imo) more akin to this
```idris
      b : Type
      a : Type
      f : a -> b
      t : Type
      g : t -> a
      xs : List t
    --------------------------------------
    Main.example_rhs : map f (map g xs) = map (\x => f (g x)) xs

    where map is the Prelude.List.List implementation of Prelude.Functor.Functor map
```
You could also perhaps hide the full signature behind a flag or an option. I understand that for some debugging cases, it would be nice to see exactly what implementation of an interface method is selected, but in most cases this is just visual noise

**Comment by adamschoenemann** [comment]
If someone would point me in the right direction source-code wise, I'd gladly take a swing at this :)

**Comment by david-christiansen** [comment]
What's going on here is that Idris interfaces are not part of the core language. Instead, they are represented as record types, and implementations are represented as inhabitants of these record types. Then, high-level accessors are generated that have interface constraints.

For example, we might have:
````
interface Functor (f : Type -> Type) where
  map : (a -> b) -> f a -> f b

implementation Functor List where
  map f [] = []
  map f (x::xs) = f x :: map f xs
````
expanding to something like
```
data Functor : (Type -> Type) -> Type where
  MkFunctor : (f : Type -> Type) -> (map : (a : Type) -> (b : Type) -> (a -> b) -> f a -> f b) -> Functor f

map : Functor f => (a -> b) -> f a -> f b
map @{MkFunctor f mapImpl} g x = mapImpl _ _ g x

mapList : (a -> b) -> List a -> List b
mapList f [] = []
mapList f (x :: xs) = f x :: mapList f xs

listFunctor : Functor List
listFunctor = MkFunctor List (\a, b => mapList {a} {b})
```

However, the user never writes most of the names in that expanded code. In particular, they never write `MkFunctor`, `mapList`, or `listFunctor`. This means that Idris needs to invent names for them.

This happens using the `SpecialName` datatype in `TT.hs`, which has various names that are meant to be fairly unique and understandable, but that have no syntax in Idris. These are injected into the normal `Name` datatype using its `SN` constructor, so sometimes we refer to them as SNs. When an SN gets printed to the the user, a fairly verbose syntax is presently used, which is what you're seeing above. In the example, `Prelude.List.List implementation of Prelude.Functor.Functor, method map` corresponds to `mapList` in my expanded example. Internally, it's something like `SN (WhereN (ImplementationN [T.pack "Prelude.List.List"] (sNS (sUN "Functor") ["Functor", "Prelude"])) (MethodN (sUN "map)))`, and this printing is coming from the `Show` instance on `Name`.

These very long names have a real advantage: when writing proofs that involve interfaces, it can be very difficult to tell the difference between different implementations if this kind of information isn't easy to see, and @adamschoenemann's proposed message would have to be carefully disambiguated in the presence of multiple implementations occurring in the same message. But I do think it's reasonable, even though it will require a bit of testing.

Here's my proposal: adopt the message you propose for clients that are not using the IDE protocol, like the console and VIM. For clients using the IDE protocol, transmit the information out-of-band so it can be displayed using colors, tooltips, or custom messages in richer clients than the terminal. 

Here's a sketch of an implementation inside Idris:
1. The pretty printer already carries around annotations like the fully-qualified names that occur in it. So step 1 is to have it retain the full `SN` in the semantic annotations, but instead of calling `show`, have a custom printer for SNs that translates these ones into something shorter.

2. When outputting to rich clients, you're done! The IDE protocol already transmits the full versions of names. Just make sure that its behavior is unchanged in these examples, and then IDE clients can do what they want with the information. For example, Emacs will already display eldoc and tooltips.

3. When outputting to the console, there's presently a few procedures in `Output.hs` that dump pretty-printer docs to the console in `IO`. They have to be in `IO` because that allows us to support semantic colors on Windows, which doesn't do ANSI color codes. Replace the `IO` with something like `WriterT [(String, Name)] IO` so that it accumulates a mapping from displayed strings that represent SNs to the actual underlying Name. This will let you print the strings uniquely (such as by adding a 1, 2, 3 to names that would otherwise conflict) and then iterate over the accumulated list at the end to construct the "where" that you are suggesting in the example.

Does this make sense?

**Comment by adamschoenemann** [comment]
Thank you for the detailed explanation. I think it makes sense, more or less :) I'll take a swing at it later this week or the weekend, and (probably) ask more questions. But the overall plan seems reasonable. Basically, I should only have to touch `Output.hs`, accumulate a mapping from "abbreviations" to fully qualified names and print these mappings in the bottom.

**Comment by david-christiansen** [comment]
Exactly. When implementing the "print the mappings" part, it should probably use the pretty printer as well. This will make it respect the number of columns in the user's terminal and use the right highlighting for things like the names of interfaces.

**Comment by adamschoenemann** [comment]
Hmmm, I've been looking at this, and it is not as simple as I'd hoped. First of all, all the functions in `Output.hs` only deal with already printed terms, e.g. `SimpleDoc OutputAnnotation`. Manipulating these documents after their construction is not particularly easy, but it can be done for sure. But the indentation of the lines is then messed up. Furthermore, it does feel like a hack. This data structure is also not very good for doing stuff like disambiguating names and so forth, since it requires traversing the Document structure to see what namesa already there. It seems like duplicating a lot of work. Since the actual textual representation comes from the `Show` instance of `SimpleName`, called through `prettyName`, which is called through `pprintPTerm` (as far as I can see at least), it'll require threading the information if we're printing to console or to IDEMode quite far down. So, I'm not certain how best to proceed in this situation? Spend a lot of time on a fragile hack, or change the API surface for pretty printing?

**Comment by porglezomp** [comment]
Is anyone still working on this? I might try to take a swing at it if it's stalled.

**Comment by porglezomp** [comment]
For whoever comes this way, there's a good workaround for it if you're implementing an interface.
Instead of writing:
```idris
module Module

implementation Interface SomeType where
  method x = ?some_implementation
```
you can write
```idris
module Module

method : ?input -> ?output
method x = ?some_implementation

implementation Interface SomeType where
  method = Module.method
```
and the partial evaluation will get you to the pretty names instead of the ugly interface qualified names.

**Comment by ahmadsalim** [comment]
@porglezomp As far as I know no one is actively working on it, and you are more than welcome to try and improve this naming issue,
